### ISSUE TITLE
`parserOptions.project` should not throw when linting a string

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#885]
```js
const linter = new eslint.CLIEngine({
	useEslintrc: false,
	configFile: // ...
});

linter.executeOnText('const foo: number = 5;');
```

**Expected Result**

I expected the above to work.

**Actual Result**

It throws an error:

```
Parsing error: If "parserOptions.project" has been set for @typescript-eslint/parser, <text> must be included in at least one of the projects provided.
```

But it's impossible to do what the error message says, as there's no file.

Same applies if you use the `--stdin` option for ESLint.

**Additional Info**

It should not throw when the filename is `<text>`.

I lint a string to test my ESLint config: https://github.com/xojs/eslint-config-xo-typescript/blob/ed789e3243680312a2e20edffdb13dd536b0bc46/test/test.js

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.0.0` |
| `@typescript-eslint/parser`        | `2.0.0` |
| `TypeScript`                       | `3.5.3` |
| `ESLint`                           | `6.2.0` |
| `node`                             | `10.16.0` |
| `npm`                              | `6.9.0` |

**Comment by bradzacher** [comment]
The reason that it's failing is because now files that are not included within your tsconfig will not be parsed up-front. We throw errors for any files that aren't parsed up front, because when we attempt to get the existing typescript program for those files, nothing exists.

We can't special case `<text>` without applying a _completely_ different code path to the normal parse so that it goes back to the old method of constructing a brand new program tree for the file.

---

If you pass a second argument to `executeOnText`, then you can pass a filename (i.e. one that already exists in your project).

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d14d57cea21e9c6ce6a8a9bad8e862f7e1101e1b/types/eslint/index.d.ts#L483

The reason that this will work is because we accept file contents from eslint so that it can apply fixes in memory. I.e. by passing in a real filename, it will override the existing file contents, and trigger a parse of the new contents.

---

I'm not sure if we should add a special case for this (very rare) use case of linting without a valid filename.

**Comment by pleunv** [comment]
I also seem to be getting this for any new ts file I add to the project. Is this normal behavior?

**Comment by sindresorhus** [comment]
> I'm not sure if we should add a special case for this (very rare) use case of linting without a valid filename.

It's common to use this (stdin, editor plugins, testing, etc). It's just that usually you would indeed supply a filename too. However, note that ESLint supports `eslint --stdin`, and not everyone would also use the `--stdin-filename` flag.

I'm ok if you deem this as "won't fix", but could be good to maybe improve the error message in such case or at least document the limitation somewhere. It was not clear to me at all that I need to also pass the filename.

**Comment by bradzacher** [comment]
@pleunv - that is unrelated to this issue. See #890

---

> It's common to use this (stdin, editor plugins, testing, etc). *It's just that usually you would indeed supply a filename too.*

That's kind of my point - the rare case is parsing without a filename at all. If you use the eslint API _and_ provide a filename - it'll all work as expected.

Unlike the standard parsers (espree, babel-eslint, etc), our parser (when provided with the `project` option) more or less _requires_ access to the file system, because the typescript compiler takes over.
In order to provide type information, there needs to be a relationship between a tsconfig file and the code being parsed.

So this will fall within the "won't fix" category.

Happy to accept a PR to improve the error message for this specific case!
I just made some updates to the error "detector" code yesterday, so it should be pretty trivial to add.
https://github.com/typescript-eslint/typescript-eslint/blob/4a29098c5b2484c144f48594d6a62ec342a5cf30/packages/typescript-estree/src/parser.ts#L93-L127

**Comment by Hermholtz** [comment]
Got the same when trying to use eslint for an Angular project. Got these packages:
```
    "eslint": "^6.8.0",
    "eslint-plugin-angular": "^4.0.1",
    "angular-tslint-rules": "^1.20.3",
    "babel-eslint": "^10.1.0",
    "@typescript-eslint/eslint-plugin": "^2.21.0",
    "@typescript-eslint/eslint-plugin-tslint": "^2.21.0",
    "@typescript-eslint/parser": "^2.21.0",
```
and this result on the first character in first line of any Angular component file:
```
Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: apps\myapp\src\components\my.component.html.
The extension for the file (.html) is non-standard. You should add "parserOptions.extraFileExtensions" to your config.
```

How do I exclude Angular components from ESLint?

**Comment by bradzacher** [comment]
@Hermholtz - your issue is unrelated to this issue.
please file a new one

**Comment by Hermholtz** [comment]
For which package?
```
@typescript-eslint/eslint-plugin-tslint
@typescript-eslint/eslint-plugin
@typescript-eslint/parser
@typescript-eslint/experimental-utils
@typescript-eslint/typescript-estree
```

**Comment by bradzacher** [comment]
`typescript-estree`. Note that it doesn't hugely matter, as I can change it after creation if it's wrong.

**Comment by radiosilence** [comment]
Is there a way to have it simply ignore these files? I ask because sometimes `eslint` is called with a list of staged files - if one of these includes a non-project file, it would be nice to be able simply skip them instead of failing.

**Comment by bradzacher** [comment]
No. It's impossible for us to ignore files.

ESLint controls and coordinates the lint run. If you tell it to lint files X, Y, and Z, then it'll tell us to parse X, Y, and X.
We then have two choices: succeed or fail.

If we succeed, we have to succeed with the correct data, or else this can cause unexpected behaviour when eslint subsequently passes the result to the configures lint rules.

So we really can only choose to fail.

If you have files that you don't want to be linted, you should be using eslint ignore files, or you should be using eslint's overrides config to disable our parser for those files.

**Comment by bradzacher** [comment]
As per https://github.com/typescript-eslint/typescript-eslint/issues/885#issuecomment-523083616, closing as wont fix, as we can't fix it.
